File: 资料/算法学习资料/数据结构专题/program/program5.pas
const
	maxn = 5011;	
var
	n, m: longint;
	data: array [0..maxn] of array [0..maxn] of boolean;
	seed: longint;
	
function next_rand():longint;
const
	P = 1000000007;
	Q = 83978833;
	R = 8523467;
begin
	seed := (int64(Q) * seed mod P * seed + R) mod P;
	exit(seed)
end;

procedure generate_input();
var
	i, j: longint;
begin
	read(n, m);
	for i := 0 to n - 1 do
		for j := 0 to m - 1 do
			data[i][j] := ((next_rand() mod 8) > 0)
end;

function check(x1, y1, x2, y2: longint):boolean;
var
	i, j: longint;
	flag: boolean;
begin
	flag := true;
	for i := x1 to x2 do
		for j := y1 to y2 do
			if not data[i][j] then
				flag := false;
	exit(flag)
end;

function count3():int64;
var
	i, j, k, l: longint;
	ans: int64;
begin
	ans := 0;
	for i := 0 to n - 1 do
		for j := 0 to m - 1 do
			for k := i to n - 1 do
				for l := j to m - 1 do
					if check(i, j, k, l) then
						inc(ans);
	exit(ans)
end;

procedure main();
var
	i: longint;
begin
	read(seed);
	for i := 0 to 9 do begin
		generate_input;
		writeln(count3);
	end
end;

begin
	main;
end.

File: 资料/算法学习资料/数据结构专题/program/program5.c
#include <stdio.h>

#define N 5011
int n, m;
char data[N][N];

int seed;
int next_rand(){
	static const int P = 1000000007, Q = 83978833, R = 8523467;
	return seed = ((long long)Q * seed % P * seed + R) % P;
}

void generate_input(){
	int i, j;
	scanf("%d%d", &n, &m);
	for(i = 0; i < n; i++)
		for(j = 0; j < m; j++)
			data[i][j] = ((next_rand() % 8) > 0 ? 1 : 0);
}

int check(int x1, int y1, int x2, int y2){
	int i, j, flag = 1;
	for(i = x1; i <= x2; i++)
		for(j = y1; j <= y2; j++)
			if(data[i][j] == 0)
				flag = 0;
	return flag;
}

long long count3(){
	int i, j, k, l;
	long long ans = 0;
	for(i = 0; i < n; i++)
		for(j = 0; j < m; j++)
			for(k = i; k < n; k++)
				for(l = j; l < m; l++)
					if(check(i, j, k, l))
						ans++;
	return ans;
}

int main(){
	int i;
	scanf("%d", &seed);
	for(i = 0; i < 10; i++){
		generate_input();
		printf("%lld\n", count3());
	}

	return 0;
}
